File: forge/src/main/java/dev/pandasystems/pandalib/forge/platform/RegistrationHelperImpl.java
/*
 * Copyright (C) 2025 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.pandasystems.pandalib.forge.platform;

import dev.pandasystems.pandalib.platform.services.RegistrationHelper;
import dev.pandasystems.pandalib.registry.DeferredObject;
import dev.pandasystems.pandalib.registry.IdentifiableResourceReloadListener;
import net.minecraft.core.MappedRegistry;
import net.minecraft.core.RegistrationInfo;
import net.minecraft.core.Registry;
import net.minecraft.core.WritableRegistry;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceKey;
import net.minecraft.server.packs.PackType;
import net.minecraftforge.client.event.RegisterClientReloadListenersEvent;
import net.minecraftforge.event.AddReloadListenerEvent;
import net.minecraftforge.registries.RegisterEvent;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

public class RegistrationHelperImpl implements RegistrationHelper {
	private final Map<ResourceKey<? extends Registry<?>>, PendingRegistries<?>> pendingRegistries = new HashMap<>();
	private final List<Registry<?>> pendingRegistryTypes = new ArrayList<>();
	private final List<IdentifiableResourceReloadListener> serverDataReloadListeners = new ArrayList<>();
	private final List<IdentifiableResourceReloadListener> clientDataReloadListeners = new ArrayList<>();

	@Override
	@SuppressWarnings("unchecked")
	public <T> void register(DeferredObject<? extends T> deferredObject, Supplier<? extends T> supplier) {
		PendingRegistries<T> pending = (PendingRegistries<T>) pendingRegistries
				.computeIfAbsent(deferredObject.getRegistryKey(), k ->
						new PendingRegistries<>((ResourceKey<? extends Registry<T>>) deferredObject.getRegistryKey()));
		pending.add(deferredObject, supplier);
	}

	@Override
	public <T> void registerNewRegistry(Registry<T> registry) {
		pendingRegistryTypes.add(registry);
	}

	@Override
	public void registerReloadListener(PackType packType, IdentifiableResourceReloadListener listener) {
		if (packType == PackType.SERVER_DATA) {
			serverDataReloadListeners.add(listener);
		} else {
			clientDataReloadListeners.add(listener);
		}
	}

	public void registerEvent(RegisterEvent event) {
		pendingRegistries.values().forEach(pending -> pending.register(event));
	}

	@SuppressWarnings({"deprecation", "unchecked", "rawtypes"})
	public void registerNewRegistries() {
		if (BuiltInRegistries.REGISTRY instanceof MappedRegistry<?> rootRegistry)
			rootRegistry.unfreeze();

		for (Registry<?> registry : pendingRegistryTypes) {
			((WritableRegistry) BuiltInRegistries.REGISTRY).register(registry.key(), registry, RegistrationInfo.BUILT_IN);
		}

		if (BuiltInRegistries.REGISTRY instanceof MappedRegistry<?> rootRegistry)
			rootRegistry.freeze();
	}

	public void addClientReloadListenerEvent(RegisterClientReloadListenersEvent event) {
		for (IdentifiableResourceReloadListener listener : clientDataReloadListeners) {
			event.registerReloadListener(listener);
		}
	}

	public void addServerReloadListenerEvent(AddReloadListenerEvent event) {
		for (IdentifiableResourceReloadListener listener : serverDataReloadListeners) {
			event.addListener(listener);
		}
	}

	private static class PendingRegistries<T> {
		private final ResourceKey<? extends Registry<T>> registryKey;

		private final Map<DeferredObject<? extends T>, Supplier<? extends T>> entries = new HashMap<>();

		public PendingRegistries(ResourceKey<? extends Registry<T>> registryKey) {
			this.registryKey = registryKey;
		}

		public void add(DeferredObject<? extends T> deferredObject, Supplier<? extends T> objectSupplier) {
			entries.put(deferredObject, objectSupplier);
		}

		public void register(RegisterEvent event) {
			entries.forEach((deferredObject, supplier) -> {
				event.register(registryKey, deferredObject.getId(), supplier::get);
				deferredObject.bind(false);
			});
		}
	}
}

File: neoforge/src/main/kotlin/platform/RegistrationHelperImpl.kt
/*
 * Copyright (c) 2025. Oliver Froberg
 *
 * This code is licensed under the GNU Lesser General Public License v3.0
 * See: https://www.gnu.org/licenses/lgpl-3.0-standalone.html
 */

package dev.pandasystems.pandalib.neoforge.platform

import dev.pandasystems.pandalib.core.platform.RegistrationHelper
import dev.pandasystems.pandalib.api.registry.deferred.DeferredObject
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.client.renderer.entity.EntityRendererProvider
import net.minecraft.core.Registry
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.packs.PackType
import net.minecraft.server.packs.resources.PreparableReloadListener
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityType
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.neoforged.neoforge.client.event.AddClientReloadListenersEvent
import net.neoforged.neoforge.client.event.EntityRenderersEvent
import net.neoforged.neoforge.event.AddServerReloadListenersEvent
import net.neoforged.neoforge.registries.NewRegistryEvent
import net.neoforged.neoforge.registries.RegisterEvent
import java.util.function.Consumer
import java.util.function.Supplier

class RegistrationHelperImpl : RegistrationHelper {
	private val pendingRegistries: MutableMap<ResourceKey<out Registry<*>>, PendingRegistries<*>> = mutableMapOf()
	private val pendingRegistryTypes: MutableList<Registry<*>> = mutableListOf()
	private val serverDataReloadListeners = mutableListOf<Pair<ResourceLocation, PreparableReloadListener>>()
	private val clientDataReloadListeners = mutableListOf<Pair<ResourceLocation, PreparableReloadListener>>()

	override fun <T> register(deferredObject: DeferredObject<out T>, supplier: Supplier<out T>) {
		@Suppress("UNCHECKED_CAST")
		val pending: PendingRegistries<T> = pendingRegistries
			.computeIfAbsent(deferredObject.key.registryKey()) { k: ResourceKey<out Registry<*>> -> PendingRegistries(deferredObject.key.registryKey()) } as PendingRegistries<T>
		pending.add(deferredObject, supplier)
	}

	override fun <T> registerNewRegistry(registry: Registry<T>) {
		pendingRegistryTypes.add(registry)
	}

	override fun registerReloadListener(
		packType: PackType,
		listener: PreparableReloadListener,
		id: ResourceLocation,
		dependencies: Collection<ResourceLocation>
	) {
		if (packType == PackType.SERVER_DATA) {
			serverDataReloadListeners += id to listener
		} else {
			clientDataReloadListeners += id to listener
		}
	}

	fun registerEvent(event: RegisterEvent) {
		pendingRegistries.values.forEach(Consumer { pending: PendingRegistries<*> -> pending.register(event) })
	}

	fun registerNewRegistryEvent(event: NewRegistryEvent) {
		pendingRegistryTypes.forEach(Consumer { registry: Registry<*> -> event.register(registry) })
	}

	fun addServerReloadListenerEvent(event: AddServerReloadListenersEvent) {
		serverDataReloadListeners.forEach(Consumer { (id, listener) -> event.addListener(id, listener) })
	}

	fun addClientReloadListenerEvent(event: AddClientReloadListenersEvent) {
		clientDataReloadListeners.forEach(Consumer { (id, listener) -> event.addListener(id, listener) })
	}

	private class PendingRegistries<T>(private val registryKey: ResourceKey<out Registry<T>>) {
		private val entries: MutableMap<DeferredObject<out T>, Supplier<out T>> = mutableMapOf<DeferredObject<out T>, Supplier<out T>>()

		fun add(deferredObject: DeferredObject<out T>, objectSupplier: Supplier<out T>) {
			entries.put(deferredObject, objectSupplier)
		}

		fun register(event: RegisterEvent) {
			entries.forEach { (deferredObject: DeferredObject<out T>, supplier: Supplier<out T>) ->
				event.register(registryKey, deferredObject.id) { supplier.get() }
				deferredObject.bind<Any>(false)
			}
		}
	}

	val entityRendererProviders = mutableMapOf<EntityType<*>, EntityRendererProvider<*>>()
	val blockEntityRendererProviders = mutableMapOf<BlockEntityType<*>, BlockEntityRendererProvider<*>>()

	override fun <R : Entity> registerEntityRenderer(
		type: EntityType<R>,
		provider: EntityRendererProvider<R>
	) {
		entityRendererProviders.put(type, provider)
	}

	override fun <R : BlockEntity> registerBlockEntityRenderer(
		type: BlockEntityType<R>,
		provider: BlockEntityRendererProvider<R>
	) {
		blockEntityRendererProviders.put(type, provider)
	}

	@Suppress("UNCHECKED_CAST")
	fun onEntityRendererRegistryEvent(event: EntityRenderersEvent.RegisterRenderers) {
		entityRendererProviders.forEach { event.registerEntityRenderer(it.key, it.value as EntityRendererProvider<in Entity>) }
		blockEntityRendererProviders.forEach { event.registerBlockEntityRenderer(it.key, it.value as BlockEntityRendererProvider<in BlockEntity>) }
	}
}
